src/validationinterface.cpp

```diff
@@ -180,3 +199,14 @@ void CMainSignals::BlockChecked(const CBlock& block, const CValidationState& sta
 void CMainSignals::NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock> &block) {
     m_internals->NewPoWValidBlock(pindex, block);
 }
+
+void CMainSignals::AddToProcessQueue(std::function<void ()> func)
+{
+    m_internals->m_schedulerClient.AddToProcessQueue([func]() {
+        // Indicate that we're running in a ValidationInterface
+        // thread. This affects which functions can be called without risk of
+        // deadlock.
+        MarkRunningInQueue();
```

Comment:
Wouldn't it be simpler and more efficient to call MarkRunningInQueue once when the queue is created, instead of every time a notification added?

Reply:
Fixed. Now executing MarkRunningInQueue once whenever the `m_internals` scheduler is reset. The change makes the refactoring strictly unneeded, but I think it's still nice to have for testing and less duplication (though I'm happy to remove it too).